timelock-impl/src/main/java/com/palantir/atlasdb/timelock/AsyncTimelockResource.java

```diff
@@ -128,12 +135,52 @@ public void waitForLocks(@Suspended final AsyncResponse response, WaitForLocksRe
         return timelock.refreshLockLeases(tokens);
     }
 
+    @POST
+    @Path("leasable-refresh-locks")
+    public LeasableRefreshLockResponse leasableRefreshLockLeases(Set<LockToken> tokens) {
+        return timelock.leasableRefreshLockLeases(tokens);
+    }
+
+    @POST
+    @Path("leasable-lock")
+    public void leasableLock(@Suspended final AsyncResponse response, LockRequest request) {
+        IdentifiedTime identifiedStartTime = timelock.identifiedTime();
```

Comment:
recently you made some perf improvements specifically around avoiding multiple calls to the AsyncTimelockService, because you have to do a leadership check every time. This feels like it avoids that :P

Reply:
I also don't think this is the right time to take this time; at a minimum what if it takes a few seconds to actually wait for the locks to be released. I expect that this time should be calculated as part of constructing the HeldLocksCollection and merely accessed here? Might need some refactoring, maybe split the server side changes into a separate PR since this one is already big? Also, because they have to be released separately this PR cannot be merged as-is; it requires the server side changes to be merged, released internally, and then a version number updated with the client side changes; I know you know this but just wanted to make more argument for server side changes being released separately.